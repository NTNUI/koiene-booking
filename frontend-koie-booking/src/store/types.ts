// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
