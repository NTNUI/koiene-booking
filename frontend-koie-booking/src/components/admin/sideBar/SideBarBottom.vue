<template>
  <div>
    <v-list-item
      v-for="(item, index) in views"
      :id="'sideBar' + item.id"
      :key="'sideBar' + item.id"
      link
      @click="updateViewFn(index)"
    >
      <v-list-item-icon>
        <v-icon :v-text="item.icon"></v-icon>
      </v-list-item-icon>
      <v-list-item-content>
        <v-list-item-title>{{ item.title }}</v-list-item-title>
      </v-list-item-content>
    </v-list-item>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import adminViews from '@/components/admin/AdminViews';

export default Vue.extend({
  name: 'SideBarBottom',
  inject: ['updateView'],
  data() {
    return {
      views: adminViews,
      updateViewFn: (this as any).updateView as (index: string) => void // Little type hack.
    };
  }
});
</script>

<style scoped></style>
