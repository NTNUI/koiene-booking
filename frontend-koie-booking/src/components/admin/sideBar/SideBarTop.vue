<template>
  <v-list-item two-line>
    <v-list-item-avatar>
      <img src="https://ntnui.no/wp-content/uploads/2018/04/ntnui-logo-liten.png" />
    </v-list-item-avatar>
    <v-list-item-content>
      <v-list-item-title>Admin</v-list-item-title>
      <v-list-item-subtitle>Koie-styret</v-list-item-subtitle>
    </v-list-item-content>
  </v-list-item>
</template>

<script>
import Vue from 'vue';

export default Vue.extend({
  name: 'SideBarTop',
});
</script>

<style scoped></style>
