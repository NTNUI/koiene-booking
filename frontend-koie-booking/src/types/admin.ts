import { Component } from 'vue';

export interface AdminView {
  id: String; // Used to reference the buttons in SideBar
  title: String;
  icon: String;
  component: Component;
}
