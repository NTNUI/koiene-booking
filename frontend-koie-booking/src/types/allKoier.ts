export interface AllKoierData {
  koieTitle: string;
}
