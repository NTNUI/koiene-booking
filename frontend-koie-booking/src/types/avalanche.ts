import { TranslateResult } from 'vue-i18n';

export interface AvalancheData {
  result: string;
  levelFirstDay: string;
  levelSecondDay: string;
  levelThirdDay: string;
  firstDay: string;
  secondDay: string;
  thirdDay: string;
  fromDate: string;
  nextDate: string;
  lastDate: string;
  warningTypes: Array<TranslateResult>;
  colors: Array<string>;
}
