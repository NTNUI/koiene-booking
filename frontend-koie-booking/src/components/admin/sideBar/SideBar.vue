<template>
  <v-card>
    <v-navigation-drawer id="sideBarDrawer" v-model="drawer" :color="$scssVars.globalColorPrimary" permanent dark app>
      <v-list>
        <SideBarTop />
        <SideBarBottom />
      </v-list>
    </v-navigation-drawer>
  </v-card>
</template>

<script lang="ts">
import Vue from 'vue';
import SideBarTop from '@/components/admin/sideBar/SideBarTop.vue';
import SideBarBottom from '@/components/admin/sideBar/SideBarBottom.vue';

export default Vue.extend({
  name: 'SideBar',
  components: { SideBarBottom, SideBarTop },
  props: {
    currentView: {
      type: Number,
      default: 0
    }
  },
  data() {
    return {
      drawer: this.currentView
    };
  },
  watch: {
    currentView(newVal: number) {
      this.drawer = newVal;
    }
  }
});
</script>

<style lang="scss">
#sideBarDrawer {
  border-top-left-radius: 0 !important;
  border-top-right-radius: 0 !important;
}
</style>
