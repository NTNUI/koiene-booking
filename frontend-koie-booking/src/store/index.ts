import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
