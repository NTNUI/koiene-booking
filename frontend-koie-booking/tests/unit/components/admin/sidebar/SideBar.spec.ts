import Vue, { VueConstructor } from 'vue';
import Vuetify, { Vuetify as VuetifyType } from 'vuetify';
import Vuex, { Store } from 'vuex';
import { storeConfig } from '@/store';
import { cloneDeep } from 'lodash';
import scssVars from '@/styles/variables.scss';
import mockAxios from 'jest-mock-axios';

Vue.use(Vuetify);

// Utilities
import { mount, createLocalVue, Wrapper } from '@vue/test-utils';

// Components or views
import { RootState } from '@/store/types';
import SideBarBottom from '@/components/admin/sideBar/SideBarBottom.vue';
import SideBarTop from '@/components/admin/sideBar/SideBarTop.vue';
import SideBar from '@/components/admin/sideBar/SideBar.vue';

describe('Component SideBar.vue', () => {
  let localVue: VueConstructor<Vue>;
  let vuetify: VuetifyType;
  let store: Store<RootState>;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(Vuetify);
    localVue.use(Vuex);
    store = new Vuex.Store(cloneDeep(storeConfig));
    localVue.prototype.$scssVars = scssVars;
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it('Matches snapshot', () => {
    // Arrange
    const wrapper = mount(SideBar, {
      localVue,
      vuetify,
      store,
      provide: {
        getCurrentView: () => null,
        setCurrentView: (index: number) => null
      }
    });

    // Assert
    expect(wrapper).toMatchSnapshot();
  });

  // TODO: It's crashing when ran in parallell, needs to be investigated
  it('Renders SideBarTop', () => {
    // Arrange
    const wrapper = mount(SideBar, {
      localVue,
      vuetify,
      store,
      provide: {
        getCurrentView: () => null,
        setCurrentView: (index: number) => null
      }
    });

    expect(wrapper.contains(SideBarTop)).toBe(true);
  });

  it('Renders SideBarBottom', () => {
    // Arrange
    const wrapper = mount(SideBar, {
      localVue,
      vuetify,
      store,
      provide: {
        getCurrentView: () => null,
        setCurrentView: (index: number) => null
      }
    });

    expect(wrapper.contains(SideBarBottom)).toBe(true);
  });
});
