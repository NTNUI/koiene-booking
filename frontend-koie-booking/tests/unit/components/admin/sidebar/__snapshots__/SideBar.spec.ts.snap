// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component SideBar.vue Matches snapshot 1`] = `
<div class="v-card v-sheet theme--light">
  <nav class="v-navigation-drawer v-navigation-drawer--fixed v-navigation-drawer--open theme--light" style="height: 100vh; top: 0px; max-height: calc(100% - 0px); transform: translateX(0%); width: 256px;" id="sideBarDrawer">
    <div class="v-navigation-drawer__image">
      <div class="v-responsive v-image" style="height: 100%; width: 100%;">
        <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;" name="fade-transition" mode="in-out"></div>
        <div class="v-responsive__content"></div>
      </div>
    </div>
    <div class="v-navigation-drawer__content">
      <div class="v-list v-sheet v-sheet--tile theme--light" style="height: 100vh;">
        <div tabindex="-1" class="v-list-item v-list-item--two-line theme--light">
          <div class="v-avatar v-list-item__avatar" style="height: 40px; min-width: 40px; width: 40px;"><img src="https://ntnui.no/wp-content/uploads/2018/04/ntnui-logo-liten.png"></div>
          <div class="v-list-item__content">
            <div class="v-list-item__title">Admin</div>
            <div class="v-list-item__subtitle">Koie-styret</div>
          </div>
        </div>
        <div>
          <div tabindex="0" id="sideBarallCabins" class="v-list-item v-list-item v-list-item--link theme--light">
            <div class="v-list-item__icon"><i aria-hidden="true" v-text="calendar-month" class="v-icon notranslate material-icons theme--light"></i></div>
            <div class="v-list-item__content">
              <div class="v-list-item__title">Alle koier</div>
            </div>
          </div>
          <div tabindex="0" id="sideBarreports" class="v-list-item v-list-item v-list-item--link theme--light">
            <div class="v-list-item__icon"><i aria-hidden="true" v-text="file-document-outline" class="v-icon notranslate material-icons theme--light"></i></div>
            <div class="v-list-item__content">
              <div class="v-list-item__title">Rapporter</div>
            </div>
          </div>
          <div tabindex="0" id="sideBarkeyOverview" class="v-list-item v-list-item v-list-item--link theme--light">
            <div class="v-list-item__icon"><i aria-hidden="true" v-text="key" class="v-icon notranslate material-icons theme--light"></i></div>
            <div class="v-list-item__content">
              <div class="v-list-item__title">NÃ¸kler</div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-navigation-drawer__border"></div>
  </nav>
</div>
`;
